topology/probes/vpp/vpp.go

```diff
@@ -103,268 +124,511 @@ func (p *Probe) getInterfaceVrfID(ch api.Channel, index uint32) int64 {
 		p.Ctx.Logger.Error(err)
 	}
 }
 
-func (p *Probe) interfacesEvents() {
-	defer p.wg.Done()
+type memifSocket struct {
+	*memif.MemifDetails
+	filename string
+}
 
-	ch, err := p.conn.NewAPIChannel()
-	if err != nil {
-		p.Ctx.Logger.Error("API channel error: ", err)
-		return
+func (p *Probe) fetchMemif(ch api.Channel) (map[uint32]*memifSocket, error) {
+	sharedMemifSockets := make(map[uint32]*memifSocket)
+
+	req := &memif.MemifSocketFilenameDump{}
+	msg := &memif.MemifSocketFilenameDetails{}
+	reqCtx := ch.SendMultiRequest(req)
+	for {
+		if lastReply, err := reqCtx.ReceiveReply(msg); lastReply {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		sharedMemifSockets[msg.SocketID] = &memifSocket{filename: interfaceName(msg.SocketFilename)}
 	}
 
-	sub, err := ch.SubscribeNotification(p.notifChan, &interfaces.SwInterfaceEvent{})
+	req2 := &memif.MemifDump{}
+	msg2 := &memif.MemifDetails{}
+	reqCtx = ch.SendMultiRequest(req2)
+	for {
+		if lastReply, err := reqCtx.ReceiveReply(msg2); lastReply {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		sharedMemifSockets[msg2.SocketID].MemifDetails = msg2
+	}
+
+	return sharedMemifSockets, nil
+}
+
+func (p *Probe) synchronize(ch api.Channel) error {
+	foundInterfaces := make(map[uint32]struct{})
+	needUpdate := make(map[uint32]struct{})
+
+	req := &interfaces.SwInterfaceDump{}
+	reqCtx := ch.SendMultiRequest(req)
+	for {
+		msg := &interfaces.SwInterfaceDetails{}
+		stop, err := reqCtx.ReceiveReply(msg)
+		if stop {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		intf, found := p.interfaceMap[msg.SwIfIndex]
+		if !found || interfaceNeedUpdate(msg, intf) {
+			needUpdate[msg.SwIfIndex] = struct{}{}
+		}
+		if found {
+			foundInterfaces[msg.SwIfIndex] = struct{}{}
+		}
+		p.interfaceMap[msg.SwIfIndex] = msg
+	}
+
+	/* Update interface metadata */
+	var sharedMemifSockets map[uint32]*memifSocket
+	for index := range needUpdate {
+		metadata := make(graph.Metadata)
+		msg := p.interfaceMap[index]
+		ifName := interfaceName(msg.InterfaceName)
+		p.Ctx.Logger.Debugf("Add/Update interface %s idx %d up/down %s", ifName, int64(msg.SwIfIndex), interfaceUpDown(msg.AdminUpDown))
+
+		var socket, id int
+		if _, err := fmt.Sscanf(string(ifName), "memif%d/%d", &socket, &id); err == nil {
+			metadata["Type"] = "memif"
+
+			if sharedMemifSockets == nil {
+				if sharedMemifSockets, err = p.fetchMemif(ch); err != nil {
+					return err
+				}
+			}
+
+			if memifSock, ok := sharedMemifSockets[uint32(socket)]; ok {
+				metadata["VPP"] = &Metadata{
+					SocketFilename: memifSock.filename,
+					ID:             int64(memifSock.ID),
+					SocketID:       int64(socket),
+					Master:         memifSock.Role == 0,
+					Mode:           memifModes[memifSock.Mode],
+					RingSize:       int64(memifSock.RingSize),
+					BufferSize:     int64(memifSock.BufferSize),
+					LinkUpDown:     memifSock.LinkUpDown == 1,
+				}
+			}
+		}
+
```

Comment:
Btw, the original Filip's PR also contained a support for AF-PACKET interfaces. If it is of an interest for you, [this is how you can tell if the given interface is AF_PACKET](https://github.com/ligato/vpp-agent/blob/master/plugins/vpp/ifplugin/vppcalls/vpp2001/dump_interface_vppcalls.go#L1010) and [this is how to obtain the name of the Linux interface to which it is attached](https://github.com/ligato/vpp-agent/blob/master/plugins/vpp/ifplugin/vppcalls/vpp2001/dump_interface_vppcalls.go#L149).